rotkehlchen/chain/evm/utils.py

```diff
@@ -0,0 +1,135 @@
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.types import LP_TOKEN_CONTRACT_ABI_NAMES
+
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+TOKEN_PROTOCOL_TO_ABI_NAME: Final[dict[str, 'LP_TOKEN_CONTRACT_ABI_NAMES']] = {
+    VELODROME_POOL_PROTOCOL: 'VELO_V2_LP',
+}
+
+
+def find_price_from_pool_contract(
+        evm_inquirer: 'EvmNodeInquirer',
+        token: EvmToken,
+        token_price_func: Callable,
+        token_price_func_args: list[Any],
+        block_identifier: BlockIdentifier,
+) -> Optional[Price]:
+    """
+    Calculate the price for an LP token the contract of which is also the contract of the
+    pool it represents. For example velodrome lp tokens. The price is calculated like this:
+    price = (Total value of liquidity pool) / (Current supply of LP tokens)
+    We need:
+    - Price of token 0
+    - Price of token 1
+    - Pooled amount of token 0
+    - Pooled amount of token 1
+    - Total supply of pool token
+    """
+    address = token.evm_address
+    abi_name = TOKEN_PROTOCOL_TO_ABI_NAME[token.protocol]  # the lp token's contract abi that will be used to query the info needed for lp token price calculation  # noqa: E501
+    contract = EvmContract(
+        address=address,
+        abi=evm_inquirer.contracts.abi(abi_name),
+        deployed_block=0,
+    )
+    methods = ('token0', 'token1', 'totalSupply', 'getReserves', 'decimals')
+    multicall_method = evm_inquirer.multicall
+    if isinstance(block_identifier, int) and block_identifier <= 7929876:
+        log.error(
+            f'No multicall contract at the block {block_identifier}. {abi_name} '
+            f'query failed. Should implement direct queries',
+        )
+        return None
+
+    try:
+        output = multicall_method(
+            require_success=True,
+            calls=[(address, contract.encode(method_name=method)) for method in methods],
+            block_identifier=block_identifier,
+        )
+    except (RemoteError, BlockchainQueryError) as e:
+        log.error(
+            f'Remote error calling multicall contract for {abi_name} '
+            f'token {token.evm_address} properties: {e!s}',
+        )
+        return None
+
+    # decode output
+    decoded = []
+    for (method_output, method_name) in zip(output, methods):
+        call_success = True
+        if multicall_method == evm_inquirer.multicall_2:
+            call_success = method_output[0]
+            call_result = method_output[1]
+        else:
+            call_result = method_output
+        if call_success and len(call_result) != 0:
+            decoded_method = contract.decode(call_result, method_name)
+            if len(decoded_method) == 1:
+                # https://github.com/PyCQA/pylint/issues/4739
+                decoded.append(decoded_method[0])
+            else:
+                decoded.append(decoded_method)
```

Comment:
can you check if this is needed when using multicall? I believe is due to the difference in multicall_2/multicall but I'm not sure

Reply:
It is needed as I saw.

Reply:
Why is it needed? When can it have more than one element here?

Reply:
The token supplies are returned as a tuple of ints. So in that case the decoded_method is a tuple and we need to add the tuple in the decoded list. later we access elements of that tuple (decoded[3][0] and decoded[3][1]:

```
        token0_supply = FVal(decoded[3][0] * 10**-token0.decimals)
        token1_supply = FVal(decoded[3][1] * 10**-token1.decimals)
```